service/tablet_allocator.cc

```diff
@@ -160,15 +239,25 @@ class load_balancer_stats_manager {
 /// The outline of the inter-node balancing algorithm is as follows:
 ///
 ///   1. Determine the set of nodes whose load should be balanced.
-///   2. Pick the least-loaded node (target)
-///   3. Keep moving tablets to the target until balance is achieved with the highest-loaded node,
-///              or we hit a limit for plan size:
-///      3.1. Pick the most-loaded node (source)
-///      3.2. Pick the most-loaded shard on the source
-///      3.3. Pick one of the candidate tablets on the source shard
-///      3.4. Evaluate collocation constraints for tablet replicas, if they pass:
-///           3.4.1 Pick the least-loaded shard on the target
-///           3.4.2 Generate a migration for the candidate tablet from the source shard to the target shard
+///   2. Divide the nodes into two sets, sources and destinations.
+///      Tablets are only moved from sources to destinations.
+///      When nodes are drained (e.g. on decommission), the drained nodes are sources and all other
+///      nodes are destinations.
+///      During free load balancing, we pick a single destination node which is the least loaded node
+///      and all other nodes are sources.
```

Comment:
Isn't this too constraining? We could fix imbalances while draining.

Instead of trying to equalize all shard loads, we should equalize delta(shard_x_load - shard_x_goal). A decommissioning node has shard_x_goal = 0. A shard X with twice the storage of shard Y has twice the goal; this can't be achieved if we equalize the load.

Reply:
An example: we decommission a node in rack A. Racks B, C have no destinations, so they're paralyzed.

Reply:
/cc @raphaelsc as it relates to load-based balancing

Reply:
Yes, but that also requires a more extensive change to the algorithm.

Reply:
We can leave it for a follow-up (perhaps after a more careful redesign)